File: c++/2610.cpp
// Link: https://leetcode.com/problems/convert-an-array-into-a-2d-array-with-conditions/description/
// Runtime: 39 ms 
// Memory: 32.03 MB

class Solution 
{
public:
    std::vector<std::vector<int>> findMatrix(const std::vector<int>& nums) 
    {
        std::unordered_map<int, std::unordered_set<int>> tableWithUsedElements;
        std::vector<std::vector<int>> arrayWithConditions;

        for (const auto& val : nums)
        {
            int i = 0;
            bool isFound = false;
            while ((i < tableWithUsedElements.size()) && (!isFound))
            {
                if (tableWithUsedElements[i].find(val) == tableWithUsedElements[i].end())
                    isFound = true;
                i++;
            }

            if (isFound)
            {
                tableWithUsedElements[i - 1].insert(val);
                arrayWithConditions[i - 1].push_back(val);
            }
            else
            {
                tableWithUsedElements[i].insert(val);
                arrayWithConditions.push_back({val});
            }
        }

        return arrayWithConditions;
    }
};

File: python/2610.py
# Link: https://leetcode.com/problems/convert-an-array-into-a-2d-array-with-conditions/description/
# Runtime: 57 ms
# Memory: 17.4 MB

class Solution:
    def findMatrix(self, nums: List[int]) -> List[List[int]]:
        table_with_used_elements = defaultdict(set)
        array_with_conditions = []

        for val in nums:
            i, is_found = 0, False
            while (i < len(table_with_used_elements)) and (not is_found):
                if val not in table_with_used_elements[i]:
                    is_found = True
                i += 1
            
            if is_found:
                table_with_used_elements[i - 1].add(val)
                array_with_conditions[i - 1].append(val)
            else:
                table_with_used_elements[i].add(val)
                array_with_conditions.append([val])
        
        return array_with_conditions
